File: Source/Aladdin.CAPI.ANSI.PKCS11.Java/src/aladdin/capi/ansi/pkcs11/cipher/DES_CBC.java
package aladdin.capi.ansi.pkcs11.cipher;
import aladdin.capi.*; 
import aladdin.capi.pkcs11.*; 
import aladdin.capi.ansi.keys.*;
import aladdin.pkcs11.*;

///////////////////////////////////////////////////////////////////////////////
// Алгоритм шифрования DES в режиме CBC
///////////////////////////////////////////////////////////////////////////////
public class DES_CBC extends aladdin.capi.pkcs11.BlockMode
{
	// параметры алгоритма
	private final CipherMode.CBC parameters;  

	// конструктор
	public DES_CBC(Applet applet, CipherMode.CBC parameters)
	{
		// указать параметры алгоритма
		super(applet, PaddingMode.NONE); this.parameters = parameters; 
        
        // проверить размер блока
        if (parameters.blockSize() != 8) throw new UnsupportedOperationException(); 
	} 
	// параметры алгоритма
    @Override public Mechanism getParameters(Session sesssion)
	{
        // параметры алгоритма
        return new Mechanism(API.CKM_DES_CBC, parameters.iv()); 
	}
    // тип ключа
    @Override public final SecretKeyFactory keyFactory() { return DES.INSTANCE; } 
	// размер ключа в байтах
	@Override public final int[] keySizes() { return new int[] {8}; }
	// размер блока
	@Override public final int blockSize() { return 8; } 

	// режим алгоритма
	@Override public CipherMode mode() { return parameters; }
} 

File: Source/Aladdin.CAPI.ANSI.PKCS11/Cipher/DES_CBC.cs
using System;
using Aladdin.PKCS11; 

namespace Aladdin.CAPI.ANSI.PKCS11.Cipher
{
	///////////////////////////////////////////////////////////////////////////////
	// Алгоритм шифрования DES в режиме CBC
	///////////////////////////////////////////////////////////////////////////////
	public class DES_CBC : CAPI.PKCS11.BlockMode
	{
		// параметры алгоритма
		private CipherMode.CBC parameters;  

		// конструктор
		public DES_CBC(CAPI.PKCS11.Applet applet, CipherMode.CBC parameters)

			// сохранить переданные параметры
			: base(applet, PaddingMode.None) { this.parameters = parameters; 
        
            // проверить размер блока
            if (parameters.BlockSize != 8) throw new NotSupportedException(); 
        }
		// параметры алгоритма
		public override Mechanism GetParameters(CAPI.PKCS11.Session sesssion)
		{
			// вернуть параметры алгоритма
		    return new Mechanism(API.CKM_DES_CBC, parameters.IV); 
		}
        // тип ключей
        public override SecretKeyFactory KeyFactory { get { return Keys.DES.Instance; }}
		// размер блока
		public override int BlockSize { get { return 8; }} 

		// режим алгоритма
		public override CipherMode Mode { get { return parameters; }}
	} 
}
